ruby_programming/advanced_ruby/blocks.md

```diff
@@ -0,0 +1,547 @@
+
+~~~ruby
+def logger
+  yield
+end
+
+logger { puts 'hello from the block' }
+
+logger do
+  p [1,2,3]
+end
+~~~
+
+The example isn't that useful but hopefully you should see how blocks allow huge flexibility in how methods act upon data. Let's see if we can make it more interesting, but still not that useful. You now get given a new requirement that users want a method that allows them to write whatever they want, and it gets printed twice to the terminal. How might you handle this requirement? Well if I told you that you can call yield as many times as you want and each time it yields to the block in the same way you'd probably have a good idea.
+
+~~~ruby
+def double_vision
+  yield
+  yield
+end
+
+double_vision { puts "How many fingers am I holding up?" }
+~~~
+
+Millennial avocados Batman. That's cool.
+
+We mentioned earlier that blocks can accept arguments. But how do you pass them to the block? No problem! Just pass them as arguments to yield and they'll be passed to the blocks parameters. If you call yield more than once you can pass a different argument each time if you wanted to.
+
+Where might this be useful? Well think about each time you use `each`. You need to yield each member of the collection to the block. Say you're writing a banking api, it will allow other banks to use your library to handle all the banking matters. One such api is a method that prints all of the transactions to a statement. For our example we'll use the terminal, but it could be a spreadsheet or anything. One issue is that different banks may want to print the transaction with different formats. We can iterate through the transactions and for each one we can yield it to a block that the caller of your method attaches to the method call. They can define how the transactions will be printed to their statement and you can focus on delivering bug free banking transactions
+
+~~~ruby
+@transactions = [10, -15, 25, 30, -24, -70, 999]
+
+def transaction_statement
+  @transactions.each do |transaction|
+    yield transaction # You just yield the transaction amount
+  end
+end
+
+transaction_statement do |transaction|
+  puts "%0.2f" % transaction # The person who calls the method can define how it is handled.
+end
+~~~
+
+If another bank wanted to print their transactions another way it's no problem, they can supply their own block.
+
+What if instead you didn't want the caller to define where the transaction for output, but just the format? Like any method call in Ruby there is a return value. So when you call yield, it returns the last executed value from the block. Using the example above if you instead moved the `puts` from the block to inside the each loop your method would be in control in where the transaction were printed. Maybe you only allow exporting to a csv for example.
```

Comment:
> define where the transaction for output

should this be define "how the transaction is outputted"? I am reviewing this quite late a night though so I could be reading this wrong 😆

Reply:
Not you mate. 😆